sdk/src/main/scala/com/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala

```diff
@@ -0,0 +1,524 @@
+package com.horizen.certificatesubmitter
+
+import akka.actor.{Actor, ActorRef, ActorSystem, Props, Timers}
+import akka.pattern.ask
+import akka.util.Timeout
+import com.horizen._
+import com.horizen.block.{MainchainBlockReference, SidechainBlock}
+import com.horizen.box.WithdrawalRequestBox
+import com.horizen.cryptolibprovider.{CryptoLibProvider, FieldElementUtils}
+import com.horizen.mainchain.api.{CertificateRequestCreator, SendCertificateRequest}
+import com.horizen.params.NetworkParams
+import com.horizen.proof.SchnorrProof
+import com.horizen.proposition.SchnorrProposition
+import com.horizen.secret.SchnorrSecret
+import com.horizen.transaction.mainchain.SidechainCreation
+import com.horizen.certificatesubmitter.AbstractCertificateSubmitter._
+import com.horizen.utils.{BytesUtils, WithdrawalEpochInfo, WithdrawalEpochUtils}
+import com.horizen.websocket.client.{MainchainNodeChannel, WebsocketErrorResponseException, WebsocketInvalidErrorMessageException}
+import sparkz.core.NodeViewHolder.CurrentView
+import sparkz.core.NodeViewHolder.ReceivableMessages.GetDataFromCurrentView
+import sparkz.core.network.NodeViewSynchronizer.ReceivableMessages.SemanticallySuccessfulModifier
+import scorex.util.ScorexLogging
+import java.io.File
+import java.util
+import java.util.Optional
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.compat.Platform.EOL
+import scala.compat.java8.OptionConverters._
+import scala.concurrent.duration.{FiniteDuration, SECONDS}
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.language.postfixOps
+import scala.util.{Failure, Random, Success, Try}
+
+/**
+ * Certificate submitter listens to the State changes and takes care of of certificate signatures managing (generation, storing and broadcasting)
+ * If `submitterEnabled` is `true`, it will try to generate and send the Certificate to MC node in case the proper amount of signatures were collected.
+ * Must be singleton.
+ */
+abstract class AbstractCertificateSubmitter (settings: SidechainSettings,
```

Comment:
I think that you move in proper direction.
Some other parts of the system can be made generic.
We also need to consider the changes from @sgonta for the new circuit.
Let's sync together.